lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/MultiTermHighlighting.java

```diff
@@ -86,85 +87,10 @@ public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
     }
 
     @Override
-    public void visitLeaf(Query query) {
-      if (query instanceof AutomatonQuery) {
-        AutomatonQuery aq = (AutomatonQuery) query;
-        if (aq.isAutomatonBinary() == false) {
-          // WildcardQuery, RegexpQuery
-          runAutomata.add(new CharacterRunAutomaton(aq.getAutomaton()) {
-            @Override
-            public String toString() {
-              return query.toString();
-            }
-          });
-        }
-        else {
-          runAutomata.add(binaryToCharRunAutomaton(aq.getAutomaton(), query.toString()));
-        }
-      }
-      else if (query instanceof FuzzyQuery) {
-        FuzzyQuery fq = (FuzzyQuery) query;
-        if (fq.getMaxEdits() == 0 || fq.getPrefixLength() >= fq.getTerm().text().length()) {
-          consumeTerms(query, fq.getTerm());
-        }
-        else {
-          runAutomata.add(new CharacterRunAutomaton(fq.toAutomaton()){
-            @Override
-            public String toString() {
-              return query.toString();
-            }
-          });
-        }
-      }
+    public void consumeTermsMatching(Query query, String field, ByteRunAutomaton automaton) {
+      runAutomata.add(automaton.asCharArrayMatcher(query.toString()));
     }
 
   }
 
-  private static CharacterRunAutomaton binaryToCharRunAutomaton(Automaton binaryAutomaton, String description) {
-    return new CharacterRunAutomaton(Automata.makeEmpty()) { // empty here is bogus just to satisfy API
-      //   TODO can we get access to the aq.compiledAutomaton.runAutomaton ?
```

Comment:
From what I see, and I'm not certain, you've succeeded in re-using the underlying AutomatonQuery's CompiledAutomaton.runAutomaton :-)